kategory-effects/src/main/kotlin/kategory/effects/data/DeferredKW.kt

```diff
@@ -0,0 +1,110 @@
+
+    fun runDeferred(): Deferred<A> =
+            thunk(coroutineContext)
+
+    fun <B> map(f: (A) -> B): DeferredKW<B> =
+            flatMap { a: A -> pure(coroutineContext, f(a)) }
+
+    fun <B> flatMap(f: (A) -> DeferredKW<B>): DeferredKW<B> =
+            DeferredKW(coroutineContext) {
+                this@DeferredKW.attempt().fold(
+                        { raiseError<B>(coroutineContext, it).runDeferred() },
+                        { f(it).runDeferred() })
+            }
+
+    companion object {
+        inline operator fun <A> invoke(coroutineContext: CoroutineContext, noinline f: suspend CoroutineScope.() -> A): DeferredKW<A> =
+                DeferredKW(coroutineContext) { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            f()
+                        }
+                    }
+                }
+
+        inline fun <A> async(coroutineContext: CoroutineContext, noinline fa: Proc<A>): DeferredKW<A> =
+                DeferredKW(coroutineContext) { context: CoroutineContext ->
+                    Platform.onceOnly<Unit> { Unit }.let {
+                        async(context, CoroutineStart.DEFAULT) {
+                            IO.async(fa).unsafeRunSync()
+                        }
+                    }
+                }
+
+        inline fun <A> pure(coroutineContext: CoroutineContext, a: A): DeferredKW<A> =
+                DeferredKW.invoke(coroutineContext, { a })
+
+        inline fun <A> raiseError(coroutineContext: CoroutineContext, t: Throwable): DeferredKW<A> =
+                DeferredKW(coroutineContext) { throw t }
+
+        fun <A, B> tailRecM(coroutineContext: CoroutineContext, a: A, f: (A) -> DeferredKW<Either<A, B>>): DeferredKW<B> {
+            tailrec fun go(a: A, f: (A) -> DeferredKW<Either<A, B>>): DeferredKW<B> {
+                val result: DeferredResult<Either<A, B>> = f(a).attempt()
+                /* FIXME(paco): KT-20075 If you remove return here tailrec stops working. Jetbrains Please. */
+                return when (result) {
+                    is Either.Left -> raiseError(coroutineContext, result.a)
+                    is Either.Right -> {
+                        val next: Either<A, B> = result.b
+                        when (next) {
+                            is Either.Left -> go(next.a, f)
+                            is Either.Right -> pure(coroutineContext, next.b)
+                        }
+                    }
+                }
+            }
+            return go(a, f)
+        }
+
+        inline fun instances(coroutineContext: CoroutineContext): DeferredKWInstances =
+                object : DeferredKWInstances {
+                    override fun CC(): CoroutineContext = coroutineContext
+                }
+
+        inline fun functor(coroutineContext: CoroutineContext): Functor<DeferredKWHK> = instances(coroutineContext)
```

Comment:
This would away once auto derived but if any needed to remain with the context as argument the return type of these factory style method should be the concrete InstanceImpl. When we use the concrete impls users can access the concrete return types without having to manually `ev()` every result obtained from using the combinators in the typeclass.